data/artifact/v0/chat_operation.go

```diff
@@ -0,0 +1,239 @@
+	defer connection.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	ctx = metadata.NewOutgoingContext(ctx, getRequestMetadata(e.SystemVariables))
+
+	res, err := artifactClient.ListMessages(ctx, &artifactPB.ListMessagesRequest{
+		NamespaceId:           inputStruct.Namespace,
+		CatalogId:             inputStruct.CatalogID,
+		ConversationId:        inputStruct.ConversationID,
+		IncludeSystemMessages: true,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("failed to list messages: %w", err)
+	}
+
+	output := RetrieveChatHistoryOutput{
+		Messages: make([]Message, 0),
+	}
+
+	output.Filter(inputStruct, res.Messages)
+
+	for res.NextPageToken != "" || (len(output.Messages) < inputStruct.MaxMessageCount && inputStruct.MaxMessageCount > 0) {
+		res, err = artifactClient.ListMessages(ctx, &artifactPB.ListMessagesRequest{
+			NamespaceId:           inputStruct.Namespace,
+			CatalogId:             inputStruct.CatalogID,
+			ConversationId:        inputStruct.ConversationID,
+			IncludeSystemMessages: true,
+			PageToken:             res.NextPageToken,
+		})
+
+		if err != nil {
+			return nil, fmt.Errorf("failed to list messages: %w", err)
+		}
+
+		output.Filter(inputStruct, res.Messages)
+	}
+
+	return base.ConvertToStructpb(output)
+
+}
+
+type WriteChatMessageInput struct {
+	Namespace      string `json:"namespace"`
+	CatalogID      string `json:"catalog-id"`
+	ConversationID string `json:"conversation-id"`
+	Content        string `json:"content"`
+	Role           string `json:"role"`
+	MessageType    string `json:"message-type"`
+}
+
+type WriteChatMessageOutput struct {
+	MessageUID  string `json:"message-uid"`
+	Role        string `json:"role"`
+	MessageType string `json:"message-type"`
+	Content     string `json:"content"`
+	CreateTime  string `json:"create-time"`
+	UpdateTime  string `json:"update-time"`
+}
+
+func (in *WriteChatMessageInput) Validate() error {
+	if in.Role != "" && in.Role != "user" && in.Role != "assistant" {
+		return fmt.Errorf("role must be either 'user' or 'assistant'")
+	}
+
+	if in.MessageType != "" && in.MessageType != "MESSAGE_TYPE_TEXT" {
+		return fmt.Errorf("message-type must be 'MESSAGE_TYPE_TEXT'")
+	}
+
+	return nil
+}
+
+func (e *execution) writeChatMessage(input *structpb.Struct) (*structpb.Struct, error) {
+	inputStruct := WriteChatMessageInput{}
+
+	err := base.ConvertFromStructpb(input, &inputStruct)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert input struct: %w", err)
+	}
+
+	err = inputStruct.Validate()
+
+	if err != nil {
+		return nil, fmt.Errorf("invalid input: %w", err)
+	}
+
+	artifactClient, connection := e.client, e.connection
+	defer connection.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Minute)
+	defer cancel()
+
+	ctx = metadata.NewOutgoingContext(ctx, getRequestMetadata(e.SystemVariables))
+
+	_, err = artifactClient.CreateConversation(ctx, &artifactPB.CreateConversationRequest{
```

Comment:
please use ListConversations.
If there is no Conversation, call CreateConversation.